brainiak/fcma/fcma.py

```diff
@@ -0,0 +1,379 @@
+    """
+    comm = MPI.COMM_WORLD
+    rank = comm.Get_rank()
+    labels = []
+    raw_data = []
+    if rank==0:
+        activity_data = readActivityData(data_dir, extension, mask_file)
+        epoch_list = np.load(epoch_file) # a list of numpy array in shape (condition, nEpochs, nTRs)
+        raw_data, labels=separateEpochs(activity_data, epoch_list)
+        time1 = time.time()
+    raw_data_length = len(raw_data)
+    raw_data_length = comm.bcast(raw_data_length, root=0)
+    for i in range(raw_data_length):  # broadcast the data subject by subject to prevent size overflow
+        if rank!=0:
+            raw_data.append(None)
+        raw_data[i] = comm.bcast(raw_data[i], root=0)
+    labels = comm.bcast(labels, root=0)
+    if rank == 0:
+        time2 = time.time()
+        print('data broadcasting done, takes', time2-time1, 's')
+    return raw_data, labels
+
+class VoxelSelector:
+    """Correlation-based voxel selection component of FCMA
+
+    Parameters
+    ----------
+
+    raw_data : numpy array of array of float
+        Each array of float contains the activity data of an epoch of a subject,
+            which essentially is a nVoxels x nTRs matrix but is serialized as one dimensional vector
+        Assumption: 1. all activity data contains the same number of voxels
+                    2. the activity data has been z-scored, ready to compute correlation as matrix multiplication
+                    3. all subjects have the same number of epochs
+                    4. voxel selection is always done in the auto-correlation, i.e. raw_data correlate with themselves
+                    5. the classifier used is SVM with linear kernel
+
+    epochs_per_subj : int
+        The number of epochs of each subject
+
+    num_voxels: int
+        The number of voxels participating in the voxel selection
+
+    labels: 1D array of int
+        Describing the condition labels of the epochs, the length of labels equals the number of epochs,
+            i.e. the length of raw_data
+
+    num_folds: int
+        The number of folds to be conducted in the cross validation
+
+    voxel_unit: int, default 100
+        The number of voxel assigned to a worker each time
+    """
+    def __init__(self, raw_data, epochs_per_subj, labels, num_folds, voxel_unit=100):
+        self.raw_data = raw_data
+        self.epochs_per_subj = epochs_per_subj
+        self.num_voxels = raw_data[0].shape[1]
+        self.labels = labels
+        self.num_folds = num_folds
+        self.voxel_unit = voxel_unit
+        # use the following code since find_library doesn't work for libraries in customized directory in Linux
+        if sys.platform=='darwin':
+            extension = '.dylib'
+        elif sys.platform=='linux':
+            extension = '.so'
+        else:
+            raise RuntimeError("Unsupported operating system")
+        try:
+            self.blas_library = ctypes.cdll.LoadLibrary('libblas'+extension)
```

Comment:
I think that the best option to avoid platform issues here is to have a c wrapper calling the blas library as needed. In that way you can also call the C interface of blas. What do you think?